htlcswitch/link.go

```diff
@@ -2158,13 +2244,15 @@ func (l *channelLink) CheckHtlcForward(payHash [32]byte,
 		// As part of the returned error, we'll send our latest routing
 		// policy so the sending node obtains the most up to date data.
 
-		return l.createFailureWithUpdate(
-			func(upd *lnwire.ChannelUpdate) lnwire.FailureMessage {
-				return lnwire.NewFeeInsufficient(
-					amtToForward, *upd,
-				)
-			},
-		)
+		return &ForwardingError{
+			FailureMessage: l.createFailureWithUpdate(
+				func(upd *lnwire.ChannelUpdate) lnwire.FailureMessage {
+					return lnwire.NewFeeInsufficient(
+						amtToForward, *upd,
+					)
+				},
+			),
+		}
```

Comment:
Why wrap in `ForwardingError` if there is no FailureDetail yet assigned? Maybe we should always set `FailureDetail` with our internal failure code and decouple it completely from the wire message.

Reply:
`FailureDetail` is intended to add further granularity to `lnwire.FailureMessage`, so will have no value when there is no further information to add. The first run of this PR had a big enum for all the failure reasons, but it ended up having a lot of duplication with the existing wire message. 

Do we have need for an internal error code beyond this use case? We're not tied to the numerical value of this enum since we map the (message:detail) to a different rpc enum, so we can always change this to have an internal failure code when we need it?

Reply:
I think an internal error code could be useful beyond this use case. Also in the interpretation of failures, there is some sketchiness around failures that we generated ourselves. We use the wire message (even though the htlc never left our node) to convey an internal error.

I see the problem with duplication though. Also not sure if there is a good solution other than just adding the optional `FailureDetail` as you do now. Maybe only using internal errors and mapping them last minute to wire messages could work long-term, but I think you tried something like that already?

Reply:
> Maybe only using internal errors and mapping them last minute to wire messages could work long-term, but I think you tried something like that already?

This is a nice idea, I can see `switchErrors` being a thing :) I do worry about duplication though, specifically in the case where add channel updates to wire messages. Would have to duplicate the metadata and error one for one in a lot of cases (`FailExpiryTooSoon`/`FailIncorrectCltvExpiry` etc)

Reply:
Maybe the channel update can also be attached last minute. The duplication may be worth it, especially if we for some reason start to deviate with our wire messages. But ok, something for later if we want to do it at all.

Reply:
What is the issue with keep returning `lnwire.FailureMessage` where we don't (yet) have a `failurDetail`, and only creating the `ForwardingError`  when that is available?

Reply:
Can actually keep returning `lnwire.FailureMessage` and then just check for `ForwardingError` type when we need the extra field 👍